IdentityCore/src/webview/embeddedWebview/MSIDAADOAuthEmbeddedWebviewController.m

```diff
@@ -0,0 +1,130 @@
+
+#if TARGET_OS_IPHONE
+#import "MSIDAppExtensionUtil.h"
+#endif
+
+@implementation MSIDAADOAuthEmbeddedWebviewController
+
+- (id)initWithStartURL:(NSURL *)startURL
+                endURL:(NSURL *)endURL
+               webview:(WKWebView *)webview
+         configuration:(MSIDWebviewConfiguration *)configuration
+               context:(id<MSIDRequestContext>)context
+{
+#if TARGET_OS_IPHONE
+    // Currently Apple has a bug in iOS about WKWebview handling NSURLAuthenticationMethodClientCertificate.
+    // It swallows the challenge response rather than sending it to server.
+    // Therefore we work around the bug by using PKeyAuth for WPJ challenge in iOS
+    NSMutableDictionary *headers;
+    if (configuration.customHeaders)
+    {
+        headers = [[NSMutableDictionary alloc] initWithDictionary:configuration.customHeaders];
+    }
+    else
+    {
+        headers = [NSMutableDictionary new];
+    }
+    [headers setValue:kMSIDPKeyAuthHeaderVersion forKey:kMSIDPKeyAuthHeader];
+    
+    configuration.customHeaders = headers;
+#endif
+    
+    return [super initWithStartURL:startURL endURL:endURL webview:webview configuration:configuration context:context];
+}
+
+- (void)decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction
+                                webview:(WKWebView *)webView
+                        decisionHandler:(void (^)(WKNavigationActionPolicy))decisionHandler
+{
+    //AAD specific policy for handling navigation action
+    NSURL *requestURL = navigationAction.request.URL;
+    NSString *requestURLString = [requestURL.absoluteString lowercaseString];
+    
+    // Stop at broker
+    if ([requestURL.scheme.lowercaseString isEqualToString:@"msauth"])
+    {
+        self.complete = YES;
+        
+        NSURL *url = navigationAction.request.URL;
+        [self completeWebAuthWithURL:url];
+        
+        decisionHandler(WKNavigationActionPolicyCancel);
+        return;
+    }
+    
+    if ([requestURL.scheme.lowercaseString isEqualToString:@"browser"])
+    {
+        self.complete = YES;
+        requestURLString = [requestURLString stringByReplacingOccurrencesOfString:@"browser://" withString:@"https://"];
+        
+#if TARGET_OS_IPHONE
+        if (![MSIDAppExtensionUtil isExecutingInAppExtension])
+        {
+            [self cancel];
+            [MSIDAppExtensionUtil sharedApplicationOpenURL:[[NSURL alloc] initWithString:requestURLString]];
+        }
+        else
+        {
+            MSID_LOG_INFO(self.context, @"unable to redirect to browser from extension");
+        }
+#else
+        [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:requestURLString]];
+#endif
+        
+        decisionHandler(WKNavigationActionPolicyCancel);
+        return;
+    }
+    
+#if TARGET_OS_IPHONE
+    // check for pkeyauth challenge.
+    if ([requestURLString hasPrefix:[kMSIDPKeyAuthUrn lowercaseString]])
+    {
+        decisionHandler(WKNavigationActionPolicyCancel);
```

Comment:
just curious, why in this case, call decisionHandler first?

Reply:
because we want to cancel the current navigation request, and then load a new request with submitUrl to submit the PKeyAuth response :)